File: 02_linked_lists/01_linked_list_values/linked_list_values.js
class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
    }
}

// Approach: Itterative -> Push node value into arr until reaching tail.
// Initiate a res to empty array
// Initial current node to head
// *While current not eq null
// **Push current node value into res
// **Assign current to head.next
// Return res array

const linkedListValues = head => {
    const res = [];
    let current = head;
    while(current != null) {
        res.push(current.val);
        current = current.next
    }
    return res;
}

const a = new Node('a')
const b = new Node('b')
const c = new Node('c')
const d = new Node('d')

a.next = b
b.next = c
c.next = d

// a -> b -> c -> d
console.log(linkedListValues(a)) // -> [ 'a', 'b', 'c', 'd' ]
const x = new Node('x')
const y = new Node('y')

x.next = y

// x -> y
console.log(linkedListValues(x)) // -> [ 'x', 'y' ]

const q = new Node('q') 

// q
console.log(linkedListValues(q)) // -> [ 'q' ]
console.log(linkedListValues(null)) // -> [ ]

module.exports = {
    linkedListValues
}

File: 02_linked_lists/01_linked_list_values/linked_list_values.py
class Node:
  def __init__(self, val):
    self.val = val
    self.next = None

# Approach: Iterative -> Push node value into arr until reaching tail.
# def linked_list_values(head):
#     res = []
#     current_node = head
#     while current_node != None:
#        res.append(current_node.val)
#        current_node = current_node.next
#     return res

# Approach: Recursive -> Use a helper function to recursively pass in the value and pointer to res array until current value is null.
def linked_list_values(head):
    res = []
    insert_value(head, res)
    return res

def insert_value(head, res):
   if head is None:
      return
   res.append(head.val)
   insert_value(head.next, res)

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")

a.next = b
b.next = c
c.next = d

# a -> b -> c -> d
print(linked_list_values(a)) # -> [ 'a', 'b', 'c', 'd' ]
x = Node("x")
y = Node("y")

x.next = y

# x -> y
print(linked_list_values(x)) # -> [ 'x', 'y' ]
q = Node("q")

# q
print(linked_list_values(q)) # -> [ 'q' ]
print(linked_list_values(None)) # -> [ ]